configuration/config2/EnvironmentMgr.cpp

```diff
@@ -142,56 +171,148 @@ std::shared_ptr<EnvironmentNode> EnvironmentMgr::addNewEnvironmentNode(const std
+std::shared_ptr<EnvironmentNode> EnvironmentMgr::addNewEnvironmentNode(const std::shared_ptr<EnvironmentNode> &pParentNode, const std::shared_ptr<SchemaItem> &pCfgItem, Status &status,
+                                                                       const std::pair<std::string, std::string> &initAttribute)
 {
-    std::shared_ptr<EnvironmentNode> pNewNode;
+    std::shared_ptr<EnvironmentNode> pNewEnvNode;
 
     //
     // Create the new node and add it to the parent
-    pNewNode = std::make_shared<EnvironmentNode>(pNewCfgItem, pNewCfgItem->getProperty("name"), pParentNode);
-    pNewNode->setId(getUniqueKey());
-    pParentNode->addChild(pNewNode);
-    addPath(pNewNode);
-    pNewNode->initialize();
+    pNewEnvNode = std::make_shared<EnvironmentNode>(pCfgItem, pCfgItem->getProperty("name"), pParentNode);
+    pNewEnvNode->setId(EnvironmentMgr::getUniqueKey());
+
+    addPath(pNewEnvNode);
+    pNewEnvNode->initialize();
+    if (!initAttribute.first.empty())
+    {
+        std::shared_ptr<EnvironmentValue> pAttr = pNewEnvNode->getAttribute(initAttribute.first);
+        if (pAttr)
+        {
+            pAttr->setValue(initAttribute.second, nullptr);
+        }
+    }
+    pParentNode->addChild(pNewEnvNode);
+
+    //
+    // Send a create event now that it's been added to the environment
+    pCfgItem->findSchemaRoot()->processEvent("create", pNewEnvNode);
+    insertExtraEnvironmentData(m_pRootNode);
+
+    //
+    // Look through the children and add any that are necessary
+    std::vector<std::shared_ptr<SchemaItem>> cfgChildren;
+    pCfgItem->getChildren(cfgChildren);
+    for (auto childIt = cfgChildren.begin(); childIt != cfgChildren.end(); ++childIt)
+    {
+        int numReq = (*childIt)->getMinInstances();
+        for (int i = 0; i<numReq; ++i)
+        {
+            std::pair<std::string, std::string> empty;
+            addNewEnvironmentNode(pNewEnvNode, *childIt, status, empty);
+        }
+    }
+
+    return pNewEnvNode;
+
+    /*pParentNode->addChild(pNewEnvNode);
+    addPath(pNewEnvNode);
+    pNewEnvNode->initialize();
 
 
     //
     // Look through the children and add any that are necessary
     std::vector<std::shared_ptr<SchemaItem>> cfgChildren;
-    pNewCfgItem->getChildren(cfgChildren);
+    pCfgItem->getChildren(cfgChildren);
     for (auto childIt = cfgChildren.begin(); childIt != cfgChildren.end(); ++childIt)
     {
         int numReq = (*childIt)->getMinInstances();
         for (int i=0; i<numReq; ++i)
         {
-            addNewEnvironmentNode(pNewNode, *childIt, status);
+            addNewEnvironmentNode(pNewEnvNode, *childIt, status);
         }
     }
 
-    return pNewNode;
+
+
+    //
+    // Send a create event now that it's been added to the environment
+    pCfgItem->findSchemaRoot()->processEvent("create", pNewEnvNode);
```

Comment:
Wonder about the name of these events being "create".  Can you explain the meaning of create... when the comment above makes it sound like an "added".